opts/filters.go

```diff
@@ -0,0 +1,98 @@
+package dhtopts
+
+import (
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+)
+
+// PublicQueryFilter returns true if the peer is suspected of being publicly accessible
+func PublicQueryFilter(h host.Host, ai peer.AddrInfo) bool {
+	if len(ai.Addrs) == 0 {
+		return false
+	}
+
+	var hasPublicAddr bool
+	for _, a := range ai.Addrs {
+		if isRelayAddr(a) {
+			return false
+		}
+		if manet.IsPublicAddr(a) {
+			hasPublicAddr = true
+		}
+	}
+	return hasPublicAddr
+}
+
+// PublicRoutingTableFilter allows a peer to be added to the routing table if the connections to that peer indicate
+// that it is on a public network
+func PublicRoutingTableFilter(conns []network.Conn) bool {
+	for _, c := range conns {
+		addr := c.RemoteMultiaddr()
+		if !isRelayAddr(addr) && manet.IsPublicAddr(addr) {
+			return true
+		}
+	}
+	return false
+}
+
+// PrivateQueryFilter returns true if the peer is suspected of being accessible over a shared private network
+func PrivateQueryFilter(h host.Host, ai peer.AddrInfo) bool {
+	conns := h.Network().ConnsToPeer(ai.ID)
+	if len(conns) > 0 {
+		for _, c := range conns {
+			if manet.IsPrivateAddr(c.RemoteMultiaddr()) {
+				return true
+			}
+		}
+		return false
+	}
```

Comment:
This may be overkill in the sense that if we're already connected to the peer and we discovered them via a private DHT query then we may as well query them if we're already connected.

The approach here was to be highly defensive of our queries being polluted with non-private peers in order to keep the private DHT small and therefore make the cost of running 2 DHTs small.

Maybe we should remove this check?

Reply:
I'd remove this check. We may learn about a few undailable peers, but we won't waste too much time dialing them.

Reply:
Note: I _would_ short-circuit and, if we _have_ a private connection, we should assume that the peer is local. I just wouldn't do the inverse "we're connected over a public IP" check.